README
pidgin-blinklight:
  Lets Blink!

About
=====
This plugin for Pidgin (or purple)  will blink the ThinkLight(TM) of your shiny
new ThinkPad notebook when you get new messages. You will need the
ibm-acpi module to be installed.

INSTALL
=======
See the INSTALL file


ChangeLog
Version 0.11.1

Fix a small bug when calculating the random blink times.

-- Joachim Breitner <nomeata@debian.org> Mon, 01 Feb 2010 12:17:26 +0100

Version 0.11

Made blinking pattern dependent on the contact’s user name, for subliminal
recognition of frequent chat partners.

-- Joachim Breitner <nomeata@debian.org> Sat, 30 Jan 2010 22:41:44 +0100

Version 0.10

Also support the /sys/class/leds/asus:phone/brightness file, patch by 
Kip, thanks.

-- Joachim Breitner <nomeata@debian.org> Thu, 21 Jun 2007 17:12:19 +0100

Version 0.9

Clean up package a bit, use Pidgin-Blinklight consistently

-- Joachim Breitner <nomeata@debian.org> Thu, 21 Jun 2007 17:12:19 +0100

Version 0.8

Gaim is gone, Pidgin’s now the bird of the day. Also rename the plugin to
pidgin-blinklight.

-- Joachim Breitner <nomeata@debian.org> Fri, 11 May 2007 15:32:15 +0200

Version 0.7

Configurability, based upon a patch from Alex Pounds <alex@alexpounds.com>, Thanks!
Works now also with ASUS mleds, I hope.
Reworked automake stuff. I still have no real clue, though.

-- Joachim Breitner <nomeata@debian.org> Tue, 13 Mar 2007 09:35:09 +0100 

Version 0.6

Compability to beta4: ui_requirement NULL instead of ""
(Might work with older versions as well, if not, tell me)

-- Joachim Breitner <nomeata@debian.org> Thu, 19 Oct 2006 18:52:25 +0000

Version 0.5

Removing possible segfault reason in thinklight-fixperm

-- Joachim Breitner <nomeata@debian.org> Wed, 16 Aug 2006 12:12:38 +0000

Version 0.4

Incuded swedish translation, kudos to Daniel Nylander <po@danielnylander.se>

-- Joachim Breitner <nomeata@debian.org> Sun, 08 Jan 2006 18:51:55 +0100

Version 0.3

Change /proc file permissoins instead of running wrapper on every blink

-- Joachim Breitner <nomeata@debian.org> Thu, 17 Nov 2005 11:04:13 +0100

Version 0.2

Make pidgin-blinklight preserve the previous state of the light
(suggestion by Stefan Ott <stefan@desire.ch>)

-- Joachim Breitner <nomeata@debian.org> Mon, 19 Sep 2005 11:28:49 +0200

Version 0.1

First release

-- Joachim Breitner <nomeata@debian.org> Mon, 27 Jun 2005 08:35:54 +0200 


src/blink.h
/* © 2005-2007 Joachim Breitner
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
#ifdef HAVE_CONFIG_H
# include "../config.h"
#endif

#include <unistd.h>
#include <string.h>
#include <glib.h>

void
blinklight_startblink(const gchar *);

guint
blinklight_timeout_add (guint, GSourceFunc, gpointer);

char *
blinklight_init();


src/demoblink.c
/* © 2009 Joachim Breitner
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
# include "../config.h"
#endif

#include <glib.h>
#include <stdio.h>

#include "blink.h"


// Cheap mans timeout add function
guint timeout_interval = 0;
GSourceFunc timeout_function;
gpointer timeout_data;

guint
blinklight_timeout_add (guint interval, GSourceFunc function, gpointer data) {
	timeout_interval = interval;
	timeout_function = function;
	timeout_data = data;
	return 0;
}

int main(int argc, char **argv) {
	char *file = blinklight_init();

	if (file == NULL) {
		printf("Could not find LED file to use.\n");
		return 1;

	}
	printf("Using file %s.\n", file);

	if (argc>1) {
		blinklight_startblink(argv[1]);
	} else {
		blinklight_startblink(NULL);
	}

	while (timeout_interval > 0) {
		usleep(timeout_interval*1000);
		timeout_interval = 0;
		(*timeout_function)(timeout_data);
	}
}



src/thinklight.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define PROCFILE "/proc/acpi/ibm/light"

#define ON "on"
#define OFF "off"
#define TOGGLE "toggle"

int main(int argc, char* argv[]) {
	FILE *file;
	int ret;
	char *new_state = NULL;
	char old_state[10] = "";
	
	if (argc != 2) {
		fprintf (stderr,"Argument required: on, off or toggle\n");
		exit(EXIT_FAILURE);
	}

	if (strcmp(argv[1],ON) == 0)  new_state=ON;
	if (strcmp(argv[1],OFF) == 0) new_state=OFF;
	if (strcmp(argv[1],TOGGLE) == 0) {
		file = fopen(PROCFILE,"r");
		      if (file == NULL) exit(EXIT_FAILURE);
		ret = fscanf(file,"status: %9s",old_state);
		      if (ret != 0) exit(EXIT_FAILURE);
		ret = fclose(file);
		      if (ret != 0) exit(EXIT_FAILURE);

		if (strcmp(old_state,ON) == 0)  new_state=OFF;
		if (strcmp(old_state,OFF) == 0) new_state=ON;
	}

	if (new_state == NULL) {
		fprintf (stderr,"No new state defined\n");
		exit(EXIT_FAILURE);
	}
		
	file = fopen(PROCFILE,"w");
	if (file == NULL) exit(EXIT_FAILURE);
	fprintf(file,"%s",new_state);
	ret = fclose(file);
	if (ret != 0) exit(EXIT_FAILURE);

	return(0);
}


src/pidgin-blinklight.c
/* © 2005-2007 Joachim Breitner
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
#ifdef HAVE_CONFIG_H
# include "../config.h"
#endif

#define PURPLE_PLUGINS

#include <unistd.h>

#ifdef ENABLE_NLS
#  include <locale.h>
#  include <libintl.h>
#  define _(x) dgettext(PACKAGE, x)
#  ifdef dgettext_noop
#    define N_(String) dgettext_noop (PACKAGE, String)
#  else
#    define N_(String) (String)
#  endif
#else
#  include <locale.h>
#  define N_(String) (String)
#  define _(x) (x)
#  define ngettext(Singular, Plural, Number) ((Number == 1) ? (Singular) : (Plural))
#endif



#include <debug.h>
// #include <pidgin.h>
#include <plugin.h>
#include <conversation.h>
#include <version.h>
#include <eventloop.h>
#include <string.h>

#include "blink.h"

guint
blinklight_timeout_add (guint interval, GSourceFunc function, gpointer data) {
	return purple_timeout_add(interval, function, data);
}

static void
gt_blink(PurpleAccount *account, const gchar *sender,
         const gchar *message, int flags, gpointer data)
{
	if (purple_prefs_get_bool("/plugins/core/pidgin-blinklight/focus"))
	{
		// Blink in any case
		blinklight_startblink(sender);
	} else {
		// Only blink when not having focus
    		PurpleConversation *conv = purple_find_conversation_with_account(PURPLE_CONV_TYPE_ANY, sender, account);
	    	if (! purple_conversation_has_focus(conv)) 
		        blinklight_startblink(sender);
	}
}

static PurplePluginPrefFrame *
get_plugin_pref_frame(PurplePlugin *plugin)
{
	PurplePluginPrefFrame *frame;
	PurplePluginPref *ppref;

	frame = purple_plugin_pref_frame_new();

	ppref = purple_plugin_pref_new_with_name_and_label(
		"/plugins/core/pidgin-blinklight/focus",
		_("Blink even when conversation has the _focus.")
	);

	purple_plugin_pref_frame_add(frame, ppref);
       
	return frame;
}

static PurplePluginUiInfo prefs_info = 
{
	get_plugin_pref_frame
};

static gboolean
gt_load(PurplePlugin *plugin) {

	// Make /proc file writable
	if (!fork())
		execl("/usr/lib/pidgin-blinklight/blinklight-fixperm","blinklight-fixperm",NULL);
	

	char *chosen_file = blinklight_init();
	if (chosen_file == NULL) {
		purple_debug_info("pidgin-blinklight","no suitable file found, deactivating plugin.");
	} else {
		purple_debug_info("pidgin-blinklight","chose file %s.\n", chosen_file);
		purple_signal_connect(
			purple_conversations_get_handle(),
			"received-im-msg",
			plugin,
			PURPLE_CALLBACK(gt_blink),
			NULL
		);
	}

	// purple_debug_info("pidgin-blinklight","pidgin-blinklight has loaded\n");
	return TRUE;
}

static gboolean
gt_unload(PurplePlugin *plugin) {

	// purple_debug_info("pidgin-blinklight","pidgin-blinklight has unloaded\n");
	return TRUE;
}

static PurplePluginInfo gt_info = {
	PURPLE_PLUGIN_MAGIC,									/* plugin magic		*/
	PURPLE_MAJOR_VERSION,									/* major purple version	*/
	PURPLE_MINOR_VERSION,									/* minor purple version	*/
	PURPLE_PLUGIN_STANDARD,									/* type			*/
	NULL,											/* ui requirement	*/
	0,											/* flags		*/
	NULL,											/* dependencies		*/
	PURPLE_PRIORITY_DEFAULT,									/* priority		*/
	"gtk-nomeata-blinklight",								/* id			*/
	NULL,											/* name			*/
	VERSION,										/* version		*/
	NULL,											/* summary		*/
	NULL,											/* description		*/

	"Joachim Breitner <mail@joachim-breitner.de>",						/* author		*/
	"http://www.joachim-breitner.de/blog/",										/* homepage		*/

	gt_load,										/* load			*/
	gt_unload,										/* unload		*/
	NULL,											/* destroy		*/

	NULL,											/* ui info		*/
	NULL,											/* extra info		*/
	&prefs_info,                                                        		        /* plugin prefs         */
	NULL											/* plugin actions	*/
};

static void
init_plugin(PurplePlugin *plugin) {

	purple_prefs_add_none("/plugins/core/pidgin-blinklight");
	purple_prefs_add_bool("/plugins/core/pidgin-blinklight/focus", TRUE);

#ifdef ENABLE_NLS
	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "UTF-8");
#endif

	gt_info.name		= _("Pidgin-Blinklight");
	gt_info.summary		= _("Flickering Messages");
	gt_info.description	= _(	"Pidgin-Blinklight:\n"
					"Flashes the Blinklight upon new messages\n"
					"To use this, you need to have the ibm-acpi or asus-laptop "
					"kernel module loaded.");

	// purple_debug_info("pidgin-blinklight","pidgin-blinklight has init'ed");
}

PURPLE_INIT_PLUGIN(Pidgin-Blinklight, init_plugin, gt_info)


src/blink.c
/* © 2009 Joachim Breitner
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
#ifdef HAVE_CONFIG_H
# include "../config.h"
#endif

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <blink.h>

#define OFF    0
#define ON     1
#define TOGGLE 2


struct interface {
	char *sysfs;
	char *commands[2];
	char *scanline;
};

#define INTERFACES 3
static struct interface interfaces[] = {
	{"/proc/acpi/ibm/light", 	                {"off","on"},	"status: %9s" },
	{"/proc/acpi/asus/mled", 	                {"0","1"},	    "%9s" },
	{"/sys/class/leds/asus:phone/brightness",   {"0","1"},	    "%9s" }
};

static struct interface *interface = NULL;

typedef struct {
	int	state;
	int	time;
} blinky;

typedef struct {
	blinky *seq;
	int pos;
} blinkstate;

/* TODO: We should really allow an option to continue blinking indefinitely
         until user opens message */
blinky default_seq[] = {
	{TOGGLE,	150},
	{TOGGLE,	125},
	{TOGGLE,	150},
	{TOGGLE,	0}
};

static guint
blinklight_blink(blinkstate *bstate) {
	FILE *file;
	char *new_state = NULL;
	char old_state[10] = "";
	int ret;

	blinky *seq = &bstate->seq[bstate->pos];

	if (interface == NULL) return FALSE;

	// purple_debug_info("pidgin-blinklight","blink called with parameter: %i\n", seq->state);
	
	if (seq->state == ON)  new_state=interface->commands[ON];
        if (seq->state == OFF) new_state=interface->commands[OFF];
	if (seq->state == TOGGLE) {
 	        // purple_debug_info("pidgin-blinklight","trying to toggle\n");
                file = fopen(interface->sysfs,"r");
                      if (file == NULL) { perror ("Trying to open sysfs for reading"); return FALSE;};
                ret = fscanf(file,interface->scanline,old_state);
                      if (ret == EOF)  { perror ("Trying to read from sysfs"); return FALSE;};
                ret = fclose(file);
                      if (ret != 0) { perror ("Trying to close sysfs"); return FALSE;};
 	        // purple_debug_info("pidgin-blinklight","done reading old state %s\n", old_state);

                if (strcmp(old_state,interface->commands[ON])  == 0) new_state=interface->commands[OFF];
                if (strcmp(old_state,interface->commands[OFF]) == 0) new_state=interface->commands[ON];
        }
	
	if (new_state == NULL) {
                // purple_debug_info("pidgin-blinklight","No new state defined\n");
		return FALSE;
        }
 	// purple_debug_info("pidgin-blinklight","setting new state: %s\n", new_state);

	file = fopen(interface->sysfs,"w");
                if (file == NULL) { perror ("Trying to open sysfs for writing"); return FALSE;};
        ret = fprintf(file,"%s",new_state);
	        if (ret < 0)   { perror ("Trying to write to sysfs"); return FALSE;};
        ret = fclose(file);
                if (ret != 0) { perror ("Trying to close sysfs"); return FALSE;};

 	// purple_debug_info("pidgin-blinklight","done setting new state: %s\n", new_state);
	
	if (seq->time) { 
		bstate->pos++;
		blinklight_timeout_add(seq->time,(GSourceFunc)blinklight_blink,bstate);
	} else {
		free(bstate->seq);
		free(bstate);
	}
	return FALSE;
}

// From http://en.wikipedia.org/wiki/Jenkins_hash_function
static guint32
jenkins_one_at_a_time_hash(const gchar *key)
{
    guint32 hash = 0;
 
    while (*key){
        hash += *(key++);
        hash += (hash << 10);
        hash ^= (hash >> 6);
    }
    hash += (hash << 3);
    hash ^= (hash >> 11);
    hash += (hash << 15);
    return hash;
}


void
blinklight_startblink(const gchar *seed) {
	int length = 4;
	blinkstate *bstate = malloc(sizeof(blinkstate));
	blinky *seq ;

 	if (seed == NULL) {
		seq = malloc(sizeof(default_seq));
		memcpy(seq, default_seq, sizeof(default_seq));

	} else {
		seq = calloc(sizeof(blinky),length);
		
		// Initialize to four toggle commands
		for (int i=0; i<length; i++) {
			seq[i].state = TOGGLE;
		}

		guint32 hash = jenkins_one_at_a_time_hash(seed);
		// Set timeing based on string
		// Range: 500 + [0,1]*200 ms
		for (int i=0; i<length-1; i++) {
			// Uses last 8 bits
			seq[i].time = 50 + hash % 200;
			// So remove them
			hash = (hash >> 8); 
			// printf("Time %d: %d\n", i, seq[i].time);
		}
	}

	bstate->seq = seq;
	bstate->pos = 0;
	blinklight_blink(bstate);
}

char *
blinklight_init() {
	// figure out which interface to use
	for (int i=0; i< INTERFACES; i++) {
		if (! access(interfaces[i].sysfs, R_OK)) {
			// File exists and is readable (not checking writable because of possible race condition)
			interface = &interfaces[i];
			return interfaces[i].sysfs;
		}
		
	}
	return NULL;
}
